crates/bevy_app/src/main_schedule.rs

```diff
@@ -0,0 +1,127 @@
+pub struct First;
+
+/// The schedule that contains work required to make [`Update`] logic.
+#[derive(ScheduleLabel, Clone, Debug, PartialEq, Eq, Hash)]
+pub struct PreUpdate;
+
+/// Runs state transitions.
+#[derive(ScheduleLabel, Clone, Debug, PartialEq, Eq, Hash)]
+pub struct StateTransition;
+
+/// Runs the [`FixedUpdate`] schedule in a loop according until all relevant elapsed time has been "consumed".
+#[derive(ScheduleLabel, Clone, Debug, PartialEq, Eq, Hash)]
+pub struct FixedUpdateLoop;
+
+/// The schedule that contains systems which only run after a fixed period of time has elapsed.
+///
+/// The exclusive `run_fixed_update_schedule` system runs this schedule.
+#[derive(ScheduleLabel, Clone, Debug, PartialEq, Eq, Hash)]
+pub struct FixedUpdate;
+
+/// The schedule that contains app logic.
+#[derive(ScheduleLabel, Clone, Debug, PartialEq, Eq, Hash)]
+pub struct Update;
+
+/// The schedule that contains systems that respond to [`Update`].
+#[derive(ScheduleLabel, Clone, Debug, PartialEq, Eq, Hash)]
+pub struct PostUpdate;
+
+/// Runs last in the schedule
+#[derive(ScheduleLabel, Clone, Debug, PartialEq, Eq, Hash)]
+pub struct Last;
+
+/// Defines the schedules to be run for the [`Main`] schedule, including
+/// their order.
+#[derive(Resource, Debug)]
+pub struct MainScheduleOrder {
+    labels: Vec<Box<dyn ScheduleLabel>>,
+}
+
+impl Default for MainScheduleOrder {
+    fn default() -> Self {
+        Self {
+            labels: vec![
+                Box::new(First),
+                Box::new(PreUpdate),
+                Box::new(StateTransition),
+                Box::new(FixedUpdateLoop),
+                Box::new(Update),
+                Box::new(PostUpdate),
+                Box::new(Last),
+            ],
+        }
+    }
+}
+
+impl MainScheduleOrder {
+    /// Adds the given `schedule` after the `after` schedule
+    pub fn insert_after(&mut self, after: impl ScheduleLabel, schedule: impl ScheduleLabel) {
+        let index = self
+            .labels
+            .iter()
+            .position(|current| (**current).eq(&after))
+            .unwrap_or_else(|| panic!("Expected {after:?} to exist"));
+        self.labels.insert(index + 1, Box::new(schedule));
+    }
+}
+
+impl Main {
+    /// A system that runs the "main schedule"
+    pub fn run_main(world: &mut World, mut run_at_least_once: Local<bool>) {
+        if !*run_at_least_once {
+            world.run_schedule(PreStartup);
+            world.run_schedule(Startup);
+            world.run_schedule(PostStartup);
+            *run_at_least_once = true;
+        }
+
+        world.resource_scope(|world, order: Mut<MainScheduleOrder>| {
+            for label in &order.labels {
+                world.run_schedule_ref(&**label);
+            }
+        });
+    }
+
+    /// Initializes the [`Main`] schedule and sub-schedules on the given `app`.
+    pub fn init(app: &mut App) {
+        app.init_schedule(Main)
+            .init_schedule(PreStartup)
```

Comment:
We've moved away from this pattern elsewhere, in favor of lazy initialization. Being able to simply omit and gracefully handle empty schedules here would be nice.

Reply:
Hmmm `world.run_schedule` and `world.run_schedule_ref` panic if a schedule doesn't exist. We don't really gracefully handle this anywhere else / I'm not seeing any other instances of lazy schedule init. Things like the `OnEnter` schedules are also initialized before running them. I think this is a "good thing". If someone is manually running a schedule, I think it should be created first (for the same reason we fail systems accessing non-existent resources).

Reply:
there is an open pr for adding try_run_schedule https://github.com/bevyengine/bevy/pull/8028

Reply:
I guess we can skip the overhead of running an empty schedule unnecessarily this way, in addition to removing init code. I'm sold.

Reply:
Gotta make configure_set lazy init schedules too or this won't work well (doing that now). I guess theres also the problem of diagnostics: the app no longer has the schedules registered, so they won't show up.
For now I'll do it this way, but theres a chance we might want to walk this back for reporting reasons.

Reply:
We also need to remove the warnings here if we're embracing this pattern or we're going to get spammed constantly when we call `try_run_schedule` on non-existent schedules:
![image](https://user-images.githubusercontent.com/2694663/226065401-6099f713-60bc-4ab1-b4d2-cf9932d0ef46.png)
 
 
![image](https://user-images.githubusercontent.com/2694663/226065485-d662867f-0a38-4fa1-a2d7-6b81f36dc85d.png)

Reply:
Agreed, we should remove those.